xls/scheduling/pipeline_schedule.cc

```diff
@@ -141,6 +144,175 @@ absl::StatusOr<ScheduleCycleMap> ScheduleToMinimizeRegisters(
+      std::pair<Node*, Node*> path(src, dst);
+      result.push_back(std::make_pair(path, ccp));
+    }
+  }
+  return result;
+}
+
+using namespace operations_research;
+
+// Schedule to minimize the total pipeline registers using SDC scheduling
+// the constraint matrix is totally unimodular, this ILP problem can be solved
+// by LP.
+// Reference:
+//   - Cong, Jason, and Zhiru Zhang. "An efficient and versatile scheduling
+//   algorithm based on SDC formulation." 2006 43rd ACM/IEEE Design Automation
+//   Conference. IEEE, 2006.
+//   - Zhang, Zhiru, and Bin Liu. "SDC-based modulo scheduling for pipeline
+//   synthesis." 2013 IEEE/ACM International Conference on Computer-Aided Design
+//   (ICCAD). IEEE, 2013.
+absl::StatusOr<ScheduleCycleMap> ScheduleToMinimizeRegistersSDC(
+    FunctionBase* f, int64_t pipeline_stages,
+    const DelayEstimator& delay_estimator, sched::ScheduleBounds* bounds,
+    int64_t clock_period_ps) {
+  XLS_VLOG(3) << "ScheduleToMinimizeRegistersSDC()";
+  XLS_VLOG(3) << "  pipeline stages = " << pipeline_stages;
+  XLS_VLOG_LINES(4, f->DumpIr());
+
+  XLS_VLOG(4) << "Initial bounds:";
+  XLS_VLOG_LINES(4, bounds->ToString());
+
+  std::unique_ptr<MPSolver> solver(MPSolver::CreateSolver("SCIP"));
+  if (!solver) {
+    XLS_LOG(WARNING) << "SCIP solver unavailable.";
+    return absl::UnavailableError("SCIP solver unavailable.");
+  }
+
+  const double infinity = solver->infinity();
+
+  absl::flat_hash_map<Node*, MPVariable*>
+      node_cycle_var;  // Node's cycle after scheduling
+  absl::flat_hash_map<Node*, MPVariable*>
+      node_lifetime_var;  // Node's lifetime, since it finishes until it's
+                          // consumed by the last user
+  for (Node* node : f->nodes()) {
+    node_cycle_var[node] =
+        solver->MakeNumVar(bounds->lb(node), bounds->ub(node), node->GetName());
+    node_lifetime_var[node] = solver->MakeNumVar(
+        0.0, infinity, absl::StrFormat("lifetime_%s", node->GetName()));
+  }
+
+  for (Node* node : f->nodes()) {
+    MPVariable* const var_node = node_cycle_var.at(node);
+    MPVariable* const var_node_lifetime = node_lifetime_var.at(node);
+    for (Node* user : node->users()) {
+      MPVariable* const var_user = node_cycle_var.at(user);
+
+      // Constraint: cycle[i_user] - cycle[i] <= lifetime[i]
+      MPConstraint* const c = solver->MakeRowConstraint(-infinity, 0);
+      c->SetCoefficient(var_user, 1);
+      c->SetCoefficient(var_node, -1);
+      c->SetCoefficient(var_node_lifetime, -1);
+    }
+  }
+
+  XLS_ASSIGN_OR_RETURN(auto node_delay, ComputeNodeDelay(f, delay_estimator));
+  for (auto [path, path_delay] :
+       ComputeCriticalCombinationalPaths(f, node_delay)) {
+    if (path_delay < clock_period_ps) {
+      continue;
+    }
+    auto [src, dst] = path;
+    // Constraint: cycle[src] - cycle[dst] <= -path_delay/clock_period_ps
+    MPConstraint* const c =
+        solver->MakeRowConstraint(-infinity, -path_delay / clock_period_ps);
```

Comment:
`path_delay` and `clock_period_ps` are `int64_t`, so this will compute `-floor(path_delay / clock_period_ps)`, which will include some invalid situations. For example, if the ratio is 5.7, then that means that `cycle[dst] - cycle[src] >= 6` is the right constraint to generate, but you are generating `cycle[src] - cycle[dst] <= -5`.

When I switch over to this formulation, I get "the problem does not have an optimal solution" for some of our internal benchmarks, so something else must be wrong as well. Without that, we get schedules that do not meet timing.